test/monkeypatch.ts

```diff
@@ -0,0 +1,192 @@
+import {assert} from "chai";
+import {DetailContext, IErrorDetail, NoopContext} from "../lib/util";
+
+/**
+ * Replace the method called `name` of `cls`
+ * with a patched method that calls `replacement`.
+ *
+ * `replacement` gets two arguments:
+ * - `instance` is the instance of `cls` whose method is being called, i.e. `this`.
+ * - `original` is the original method that was patched, with all arguments bound.
+ *      You must call it with no arguments and return its result.
+ *
+ * The same method can be patched multiple times and each successive patch
+ * will call the previous patch as expected.
+ */
+const patchMethod = (
+  cls: any,
+  name: string,
+  replacement: (instance: any, original: Function) => void
+): void => {
+  const originalMethod: Function = cls.prototype[name];
+  cls.prototype[name] = function () {
+    const original = originalMethod.bind(this, ...Array.from(arguments));
+    return replacement(this, original);
+  }
+};
+
+/**
+ * In fork() and in failed(),
+ * assert that the user shouldn't have returned due to previous failure(s).
+ */
+{
```

Comment:
All this patching is pretty fancy. I am a little worried that with so much test code around the internal calls, it's hard to know if the unpatched code would pass tests (perhaps not in this diff, but in a future one made by someone less familiar).

It would be nice if we can run tests using normal unpatched library, and then run again with the patches too, to test correct usage of fork-related methods. I think it may be possible reasonably easily using Sinon.js.

Roughly, you can do:
```
const sandbox = sinon.createSandbox();
before(() => doAllPatching(sandbox));
after(() => sandbox.restore());
RUN_SAME_SUITE_AS_WITHOUT_PATCHING;

function doAllPatching(sandbox) {
   patchMethod(sandbox, NoopContext, "fork", notFailedNoop);
   ...
}
function patchMethod(sandbox, ...) {
   sandbox.replace(cls, name, function() ...);
}
```

Reply:
I considered this risk, but decided it would be pretty hard to write incorrect patches which still make the tests pass.

I can extend patchMethod a bit to assert that the replacement calls the original function exactly once and returns its value. Or I could change the API so that instead of having to call the original function and return its result, you just specify whether the patch runs before or after the original. If it runs after, you get the return value from the original which you can use, but there's no way to return you own value instead.

I'm fine with running with and without patches. But how do I run the test suite twice? I'm not managing to do it from within test_checkers.ts. Should I run the mocha command twice with and without an env var?

If I run the suite once without patches and then again with patches, surely there's no need for any sandboxes or restoring?

Reply:
Hmm, if you run mocha twice (separate processes), then no need for sandbox stuff. But what I had in mind was this:

```
describe('ts-interface-checker', suite);

describe('ts-interface-checker-with-asserts', () => {
  const sandbox = sinon.createSandbox();
  before(() => doAllPatching(sandbox));
  after(() => sandbox.restore());
  suite();
});

function suite() {
  it('should ...', () => {
  });
  ...
};
```

This way a single `mocha` command will run tests with and without the patched asserts, and the sandboxing is needed to ensure that the patching applies to just one of the runs. (BTW, I often find [sinon](https://sinonjs.org/releases/latest/) useful; it just doesn't happen to be a dependency yet. You can add it as a dev dependency, along with `@types/sinon`).

I do prefer to keep tests running without the patching. Another option is to do the patching conditionally, based on some environment variable, and run a separate mocha command from package.json, with that variable set.

Reply:
I was imagining something like `suite = describe(...); suite.run(); applyPatches(); suite.run()`. That seemed conceptually simpler than restoring a sandbox. But I guess that's not how mocha works.

Sandbox.replace doesn't work here:

> Attempts to replace an already replaced value cause an exception.

https://sinonjs.org/releases/latest/sandbox/

I quite like the logical grouping of assertions and don't want to have a single patch for each method. So I'm just restoring the prototypes manually. It's not beautiful but it works fine.

I do prefer this to an env var so that the mocha command just works on its own.